pkg/replica/diff_disk.go

```diff
@@ -110,7 +113,37 @@ func (d *diffDisk) readModifyWrite(buf []byte, offset int64) (int, error) {
 
 	copy(readBuf[offset%d.sectorSize:], buf)
 
-	return d.fullWriteAt(readBuf, readOffset)
+	n, err := d.fullWriteAt(readBuf, readOffset)
+	if err != nil {
+		return int(computeNominalWrittenBytes(int64(len(buf)), int64(n), offset, readOffset)), err
+	}
+
+	return len(buf), nil
+}
+
+// computeNominalWrittenBytes computes nominal written bytes for the input buf
+//
+//	For Example:
+//	offset > readOffset  or  offset == readOffset
+//	readOffset               readOffset
+//	0----------1---2---3     1---2---3----------.
+//	|          |  buf  |     |  buf  |          |
+//	.----------.-------.     .-------.----------.
+//	           offset        offset
+//	If wb is at point before 1 (including 1), it should return 0
+//	If wb is at point between 1 and 3 (excluding 3), it should return wb - (offset - readOffset)
+//	If wb is at point after 3 (including 3), it should return bufSize
+func computeNominalWrittenBytes(bufSize, wb, offset, readOffset int64) int64 {
+	// offset should be equal to or bigger than readOffset
+	n := wb - (offset - readOffset)
```

Comment:
> // offset should be equal to or bigger than readOffset
Should we validate this first?
```
if offset < readOffset {
}
```

Reply:
`wb` isn't very descriptive. Consider renaming to `writePos` or `writtenBytesPos` for clarity.

Similarly, `n` could be renamed to `nominal` or `nominalWritten`.

Reply:
Sounds good to me.